internal/controller/controllers/bmh_agent_controller.go

```diff
@@ -0,0 +1,499 @@
+*/
+
+package controllers
+
+import (
+	"context"
+	"encoding/json"
+	"strings"
+	"time"
+
+	bmh_v1alpha1 "github.com/metal3-io/baremetal-operator/apis/metal3.io/v1alpha1"
+	adiiov1alpha1 "github.com/openshift/assisted-service/internal/controller/api/v1alpha1"
+	"github.com/openshift/assisted-service/models"
+	"github.com/sirupsen/logrus"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+)
+
+// BMACReconciler reconciles a Agent object
+type BMACReconciler struct {
+	client.Client
+	Log    logrus.FieldLogger
+	Scheme *runtime.Scheme
+}
+
+const (
+	AGENT_BMH_LABEL                 = "adi.openshift.io/bmh"
+	BMH_AGENT_ROLE                  = "bmac.adi.openshift.io/role"
+	BMH_AGENT_HOSTNAME              = "bmac.adi.openshift.io/hostname"
+	BMH_AGENT_MACHINE_CONFIG_POOL   = "bmac.adi.openshift.io/machine-config-pool"
+	BMH_INSTALL_ENV_LABEL           = "installenvs.adi.openshift.io"
+	BMH_INSPECT_ANNOTATION          = "inspect.metal3.io"
+	BMH_HARDWARE_DETAILS_ANNOTATION = "inspect.metal3.io/hardwaredetails"
+)
+
+// reconcileResult is an interface that encapsulates the result of a Reconcile
+// call, as returned by the action corresponding to the current state.
+type reconcileResult interface {
+	Result() (reconcile.Result, error)
+	Dirty() bool
+}
+
+// reconcileComplete is a result indicating that the current reconcile has completed,
+// and there is nothing else to do.
+type reconcileComplete struct {
+	dirty bool
+}
+
+func (r reconcileComplete) Result() (result reconcile.Result, err error) {
+	return
+}
+
+func (r reconcileComplete) Dirty() bool {
+	return r.dirty
+}
+
+// reconcileRequeue is a result indicating that the reconcile should be
+// requeued.
+type reconcileRequeue struct {
+	delay time.Duration
+}
+
+func (r reconcileRequeue) Result() (result reconcile.Result, err error) {
+	result.RequeueAfter = r.delay
+	// Set Requeue true as well as RequeueAfter in case the delay is 0.
+	result.Requeue = true
+	return
+}
+
+func (r reconcileRequeue) Dirty() bool {
+	return false
+}
+
+// reconcileError is a result indicating that an error occurred while attempting
+// to advance the current reconcile, and that reconciliation should be retried.
+type reconcileError struct {
+	err error
+}
+
+func (r reconcileError) Result() (result reconcile.Result, err error) {
+	err = r.err
+	return
+}
+
+func (r reconcileError) Dirty() bool {
+	return false
+}
+
+// +kubebuilder:rbac:groups=metal3.io,resources=baremetalhosts,verbs=get;list;watch;update;patch
+
+func (r *BMACReconciler) Reconcile(req ctrl.Request) (ctrl.Result, error) {
+	ctx := context.Background()
+	bmh := &bmh_v1alpha1.BareMetalHost{}
+
+	if err := r.Get(ctx, req.NamespacedName, bmh); err != nil {
+		return reconcileComplete{}.Result()
```

Comment:
We need to requeue the request if the error is anything other than IsNotFound.

Reply:
Yes, we don't want to requeue if BMH is not found. But in case of any intermittent API errors,  we probabaly want to requeue and retry again? Eg. 
```
if err != nil {
		if errors.IsNotFound(err) {
			return reconcileComplete{}.Result()
		}
		return reconcileError{err}.Result()
}
```

Reply:
I think you're saying the same thing.You both agree that if the BMH is not found, there is no need to requeue.

@sacharya is right that if there's some other error, it is important to requeue so that this is tried again.

https://book.kubebuilder.io/cronjob-tutorial/controller-implementation.html#1-load-the-cronjob-by-name

Reply:
ah, lol! I get what @sacharya is saying now. Yes, will modify :)